framework/src/main/java/org/checkerframework/framework/flow/CFAbstractTransfer.java

```diff
@@ -366,6 +379,132 @@ public S initialStore(UnderlyingAST underlyingAST, List<LocalVariableNode> param
+    if (lambdaTree.getBodyKind() == LambdaExpressionTree.BodyKind.EXPRESSION) {
+      ExpressionTree lambdaExpression = (ExpressionTree) lambdaTree.getBody();
+      JavaExpression internalRepr = JavaExpression.fromTree(lambdaExpression);
+      return areAllMethodsPure(internalRepr, aTypeFactory)
+          && areAllArgumentsUnassignable(internalRepr);
+    } else {
+      StatementTree lambdaStatement = (StatementTree) lambdaTree.getBody();
+      return this.isStatementPure(lambdaStatement, aTypeFactory);
+    }
+  }
+
+  /**
+   * Determine whether a given statement is pure.
+   *
+   * <p>The purity of a statement depends on whether:
+   *
+   * <ol>
+   *   <li><i>All</i> methods invoked within the statement are known to be pure.
+   *   <li><i>All</i> arguments to the methods within the statement are not assignable by other code
+   * </ol>
+   *
+   * @param statementTree s statement
+   * @param aTypeFactory an annotated type factory
+   * @return true if the given statement is pure
+   */
+  private boolean isStatementPure(StatementTree statementTree, AnnotatedTypeFactory aTypeFactory) {
+    if (!(statementTree instanceof BlockTree)) {
+      return false;
+    }
+    BlockTree statementBlock = (BlockTree) statementTree;
+    for (StatementTree stmt : statementBlock.getStatements()) {
+      if (stmt instanceof ExpressionStatementTree) {
+        JavaExpression internalRepr =
+            JavaExpression.fromTree(((ExpressionStatementTree) stmt).getExpression());
+        if (!areAllMethodsPure(internalRepr, aTypeFactory)
+            || !areAllArgumentsUnassignable(internalRepr)) {
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Determine whether a sequence of method calls is pure (i.e., has no side effects and is
+   * deterministic).
+   *
+   * <p>This is solely determined by checking whether the declaration of each method call in the
+   * sequence is annotated with {@link Pure}.
+   *
+   * @param methodCallSequence the sequence of method calls to check for purity
+   * @param aTypeFactory an annotated type factory
+   * @return true if the method call sequence is comprised entirely of methods annotated with {@link
+   *     Pure}
+   */
+  private boolean areAllMethodsPure(
+      JavaExpression methodCallSequence, AnnotatedTypeFactory aTypeFactory) {
+    List<Element> methodsInvoked = JavaExpression.methodsFromMethodCall(methodCallSequence);
+    return methodsInvoked.stream()
+        .allMatch(method -> aTypeFactory.getDeclAnnotation(method, Pure.class) != null);
+  }
+
+  /**
+   * Given a method call sequence (e.g., m.foo(p1).bar().baz(p2)), determine whether all the
+   * arguments are unassignable.
+   *
+   * @param methodCallSequence the method call sequence to extract arguments from
+   * @return true if all the arguments in the method call sequence are unassignable
+   */
+  private boolean areAllArgumentsUnassignable(JavaExpression methodCallSequence) {
```

Comment:
I feel like there are better places for these methods (perhaps as static methods in `JavaExpression` at least for `areAllMethodsPure` and `areAllArgumentsUnassignable`), but I am unsure. Would like to know what you think, @mernst and @smillst.

Reply:
Sure, moving those two to `JavaExpression` sounds good.  Thanks.

Reply:
I left `areAllMethodsPure` in `CFAbstractTransfer` because moving it into `JavaExpression` would require us to introduce a dependency from dataflow to framework (for the `AnnotatedTypeFactory`), which doesn't seem ideal.

I think leaving it in is the lesser of two evils.